src/utils/proxySet.ts

```diff
@@ -0,0 +1,96 @@
+import { proxy } from 'valtio'
+
+// properies that we don't want to expose to the end-user
+interface InternalProxySet<T> extends Set<T> {
+  data: T[]
+  toJSON: object
+  hasProxy(value: T): boolean
+}
+
+export function proxySet<T>(
+  initialValues: Iterable<T> | null | undefined = []
+): Set<T> {
+  const set: InternalProxySet<T> = proxy({
+    data: Array.from(new Set(initialValues)),
+    has(value) {
+      return this.data.indexOf(value) !== -1
+    },
+    hasProxy(value) {
+      let hasProxy = false
+      if (typeof value === 'object') {
+        hasProxy = this.data.indexOf(proxy(value as any)) !== -1
+      }
+
+      return hasProxy
+    },
+    add(value) {
+      if (!this.has(value) && !this.hasProxy(value)) {
+        this.data.push(value)
+      }
+      return this
+    },
+    delete(value) {
+      if (!this.has(value) && !this.hasProxy(value)) {
+        return false
+      }
+      this.data = this.data.filter((v: T) => v !== value)
```

Comment:
Okay, we should not reassign data here. This is a subtle difference, but maybe important because how proxies work.
```js
  delete(value) {
    const index = this.data.indexOf(value)
    if (index === -1) {
      return false
    }
    this.data.splice(index, 1)
    return true
  }
```

Reply:
For my personal understanding, why we should not reassign data? This would create a new proxy under the hood, right? Is it because of that?

Reply:
Yes, reassignment will create a new proxy under the hood. With splice, we keep the same proxy without creating a new one.
<img width="559" alt="image" src="https://user-images.githubusercontent.com/490574/149507104-516bb682-72ad-4ea7-aba0-47eee203b5fa.png">
https://codesandbox.io/s/valtio-puzzle-forked-5hner